dart/lib/src/hub.dart

```diff
@@ -0,0 +1,379 @@
+import 'dart:async';
+import 'dart:collection';
+
+import 'client.dart';
+import 'protocol.dart';
+import 'scope.dart';
+import 'sentry_options.dart';
+
+typedef ScopeCallback = void Function(Scope);
+
+/// SDK API contract which combines a client and scope management
+class Hub implements HubInterface {
+  static SentryClient _getClient({SentryOptions fromOptions}) {
+    return SentryClient(
+      dsn: fromOptions.dsn,
+      environmentAttributes: fromOptions.environmentAttributes,
+      compressPayload: fromOptions.compressPayload,
+      httpClient: fromOptions.httpClient,
+      clock: fromOptions.clock,
+      uuidGenerator: fromOptions.uuidGenerator,
+    );
+  }
+
+  final ListQueue<_StackItem> _stack;
+
+  final SentryOptions _options;
+
+  factory Hub(SentryOptions options) {
+    _validateOptions(options);
+
+    return Hub._(options);
+  }
+
+  Hub._(SentryOptions options)
+      : _options = options,
+        _stack = ListQueue() {
+    _stack.add(_StackItem(_getClient(fromOptions: options), Scope(_options)));
+    _isEnabled = true;
+  }
+
+  static void _validateOptions(SentryOptions options) {
+    if (options == null) {
+      throw ArgumentError.notNull('options');
+    }
+
+    if (options.dsn == null) {
+      throw ArgumentError.notNull('options.dsn');
+    }
+  }
+
+  bool _isEnabled = false;
+
+  @override
+  bool get isEnabled => _isEnabled;
+
+  SentryId _lastEventId;
+
+  @override
+  SentryId get lastEventId => _lastEventId;
+
+  @override
+  Future<SentryId> captureEvent(Event event) async {
+    var sentryId = SentryId.empty();
+
+    if (!_isEnabled) {
+      _options.logger(
+        SeverityLevel.warning,
+        "Instance is disabled and this 'captureEvent' call is a no-op.",
+      );
+    } else if (event == null) {
+      _options.logger(
+        SeverityLevel.warning,
+        'captureEvent called with null parameter.',
+      );
+    } else {
+      final item = _stack.last;
+      if (item != null) {
+        try {
+          sentryId = await item.client.captureEvent(event: event);
+        } catch (err) {
+          /* FIXME(rxlabz) ? Event.id ?*/
+          _options.logger(
+            SeverityLevel.error,
+            'Error while capturing event with id: ${event}',
+          );
+        }
+      } else {
+        _options.logger(
+          SeverityLevel.fatal,
+          'Stack peek was null when captureEvent',
+        );
+      }
+    }
+    _lastEventId = sentryId;
+    return sentryId;
+  }
+
+  @override
+  Future<SentryId> captureException({
+    dynamic throwable,
+    dynamic stackTrace,
```

Comment:
why are we using `dynamic`? should we not use the types?

Reply:
there is no type union in Dart, and both Error and Exception are throwable, and not related. In fact you can throw anything in Dart ( but `null` which throws NullThrownError  ).

Reply:
I see, well making it dynamic works but if it could be literally anything, how we would know which fields to read the info out of the exception, stack trace, messages and so on, we'd need to think this thru, if its the case, we could offer overloads like
captureException(exception type)
captureError(error type)
capture??(any?)

I am just thinking out loud here :)

Reply:
There is no common interface for Dart `Errors`
-  a Dart `Error` has no fields https://api.dart.dev/stable/2.10.2/dart-core/Error-class.html.
- `Exception` has an optional argument ( String message ) https://api.dart.dev/stable/2.10.2/dart-core/Exception-class.html
For now we send the `exception.runtimeType` and the `exception.toString()`, whatever the exception is.
```dart
if (exception != null) {
      json['exception'] = [
        <String, dynamic>{
          'type': '${exception.runtimeType}',
          'value': '$exception',
        }
      ];
    }
```

Reply:
I see but an Error has a stack trace field https://api.dart.dev/stable/2.10.2/dart-core/Error/stackTrace.html

Reply:
we could add like this ?
```
if (exception != null) {
      json['exception'] = [
        <String, dynamic>{
          'type': '${exception.runtimeType}',
          'value': '$exception',
           if( exception is Error ) 'stackTrace': exception.stackTrace ?? ''
        }
      ];
    }
```

Reply:
or maybe ?
```dart
if (exception != null) {
      json['exception'] = [
        <String, dynamic>{
          'type': '${exception.runtimeType}',
          'value': '$exception',
           'stackTrace': '${exception.stackTrace ?? '' }' 
        }
      ];
      if( exception is Error ){
        json['stacktrace'] = exception.stackTrace ?? '';
     }
    }
```

Reply:
I'd need to dig a bit more about it, the concept of throwing anything is a bit weird to me right now, can't really tell, but yeah, it seems this conditional `if` does the job.

Reply:
I added the `exception is Error` condition

Reply:
`stackTrace` could be read in SentryClient from the `throwable ` if its an error and not taken as params in the Sentry static class/Hub

Reply:
I found the reason why Dart exception have not a stacktrace : https://groups.google.com/a/dartlang.org/g/misc/c/O1OKnYTUcoo/m/ipIE1jqHD9sJ
I think the current implementation allows to send both : 
- an exception and it's stacktrace if the user catch it and want to send it, 
- and an error with the stacktrace automatically serialized
Is it ok for you @marandaneto ?

Reply:
yeah, sounds good, but then `stacktrace` should not be mandatory right?
also, if the error has a stacktrace and the user still passed a stacktrace, which one wins? we need to think about that, but it doesn't need to be in this PR then

Reply:
only the throwable is a required argument :
```dart
Future<SentryId> captureException(
    dynamic throwable, {
    dynamic stackTrace,
  })
```
for the if a stacktrace is passed with an error, it will override the error.stacktrace

Reply:
perfect